WIP: Hca refactor validations

awong-dev opened the pull request.

akainic committed:
Refactor when validations run

Create validation logic in the model that mirrors validation logic in the view.
Initialize app data to be null values and change validations so they will return
true for null values. Then on continue click, change all remaining null values
to empty strings so validations will run at that time as expected.

awong-dev commented on _health-care/_js/components/personal-information/NameAndGeneralInfoSection.jsx:
```diff
@@ -37,9 +37,7 @@ class NameAndGeneralInfoSection extends React.Component {
-                  day={this.props.data.dateOfBirth.day}
-                  month={this.props.data.dateOfBirth.month}
-                  year={this.props.data.dateOfBirth.year}
+                  value={this.props.data.dateOfBirth}
```
why do we need to change this?

akainic replied:
1) to be consistent with how we are handling other nested object components, 2) because I thought it would avoid the console errors about required props. Can definitely revert these changes if you think that's best.

awong-dev replied:
Gotcha.

I would prefer to revert these for now.

Overall, it feels more awkward to have an object come in than individual parameters from an API perspective. It's like doing

`function renderDate(day, month, year)` vs `function renderDate(jsonOfDate)`.

Same argument could be made for some of other other more complex components (eg, FullName).

awong-dev commented:
I see where you're going...I think we should get it in right now and the maybe I can take some of the load of unittesting off of you. Thoughts?

akainic commented:
@awong-dev That sounds like a good strategy. Then once a few unit tests are built I can easily implement this for the remaining sections. Let me know what you want to do about the DateInput changes, that sort of snuck in as an initial strategy I tried, but I'm not committed to it.

awong-dev commented on _health-care/_js/components/HealthCareApp.jsx:
```diff
@@ -213,8 +213,30 @@ class HealthCareApp extends React.Component {
+      if (value === null) {
+        _.set(objectData, key, '');
+      } else if (typeof (value) === 'object') {
+        context.resetNullValues(value);
```
Yay recursion!

awong-dev commented on _health-care/_js/components/HealthCareApp.jsx:
```diff
@@ -213,8 +213,30 @@ class HealthCareApp extends React.Component {
   }
 
+  resetNullValues(objectData) {
+    const context = this;
```
eep. I don't understand the scoping of `this` to know why this code is necessary. Can we add a comment or somehow remove?

akainic replied:
Yes... I don't really understand what's going on either. I just noticed that I didn't have access to `this` once inside the function, which confuses me because I thought lexical scoping was the advantage of arrow functions?

akainic replied:
I've been digging into this for awhile and can't seem to figure out where `this` went :( For now I will add a comment to return to this later and investigate.

awong-dev commented on _health-care/_js/components/HealthCareApp.jsx:
```diff
@@ -213,8 +213,30 @@ class HealthCareApp extends React.Component {
+      }
+    });
+
+    return objectData;
```
Hmm...this API both mutates the parameter and passes it back as if it were an expression (which would imply it returns a new copy rather than changing the parameter).

Also, mutating the object while iterating over it is always a little bit weird feeling.  I would recommend using _.mapValues (https://lodash.com/docs#mapValues) to handle this. That way you don't mutate the object that you are visiting.

akainic replied:
I'm having some trouble getting `_.mapValues` to work. It seems to work fine until the function is called again recursively, and I think that's messing up the return value. Still debugging...

awong-dev commented on _health-care/_js/components/HealthCareApp.jsx:
```diff
@@ -213,8 +213,30 @@ class HealthCareApp extends React.Component {
+    this.setState({}, () => {
+      if (validationFunctions.isValidSection(sectionPath, sectionData)) {
+        hashHistory.push(this.getNextUrl());
+      }
```
If not, then it just does nothing for now...should we put a TODO here to add a flash or something?  Or is it good enough UI notification to have all the boxes go red?

akainic replied:
Good question. For now at least I will scroll to the first field with an error, and then perhaps in user testing we can get a better sense of whether that is the right kind of indication, or if we should also include a general error message or notification. Not sure of the most React way of doing this, for now I'll just do it the good old fashioned JavaScript way :)

awong-dev commented:
Back at ya.  Since we have a timezone flip, feel free to commit after addressing with a TBR=awong-dev. I'll re-review and if there are any concerns, we can address in a follow-up PR.

awong-dev committed:
Use _.mapValues to not mutate this.state.

The previous code using _.forIn() and _.set() was dangerous because it
accidentally mutated the data stored in this.state without calling
setState().

Using _.mapValues() and avoiding the mutation is cleaner and more
correct.

akainic committed:
Child, providers, and spouse not working yet

awong-dev commented on _health-care/_js/utils/validations.js:
```diff
@@ -1,57 +1,195 @@
 //        3. argument order is now based on form order... using
 function isValidAddress(street, city, country, state, zipcode) {
+  if (street !== null && city !== null && country !== null && state !== null && zipcode !== null) {
+    let n = 0;
```
You can delete all of the 
```
+    let n = 0;
+    if (street === '') n++;
 +    if (city === '') n++;
 +    if (country === '') n++;
 +    if (state === '') n++;
 +    if (zipcode === '') n++;
```

Now that you actually use all the function parameters.

akainic commented on _health-care/_js/components/HealthCareApp.jsx:
```diff
@@ -214,8 +214,30 @@ class HealthCareApp extends React.Component {
   }
 
+  resetNullValues(objectData) {
+    return _.mapValues(objectData, (value, _key) => {
```
Having another problem with `_.mapValues`: because it returns everything as an object, when it gets to `children` or `providers`, which should be an array, it turns it into an object. This means that later functionality in `GrowableTable` where it maps through the array of children or providers fails. I've tried to figure out a way for `_.mapValues` to check if the object is an array, and if so to map through the array and called itself recursively on each element in the array so as to not touch the array, but it is still returning an object for `children` and `providers`. Not sure another way around this.

awong-dev replied:
ugh.

awong-dev committed:
Unittest and fix initializeNullValues.

akainic commented on _health-care/_js/components/HealthCareApp.jsx:
```diff
@@ -7,7 +7,7 @@ import ContinueButton from './ContinueButton';
 import Nav from './Nav.jsx';
 
-import * as validationFunctions from '../utils/validations';
+import * as validations from '../utils/validations';
```
This is a much better name :)

akainic commented on _health-care/_js/utils/validations.js:
```diff
@@ -190,7 +192,20 @@ function isValidSection(completePath, sectionData) {
+function initializeNullValues(value) {
+  if (value === null) {
+    return '';
+  } else if (_.isPlainObject(value)) {
```
I'm not sure about this, but when I was testing things locally, I first noticed that arrays were returning true for `typeof (value) === 'object'`, so when I was testing checking if something was an array, I checked that first before then checking if it was (or wasn't) an object.

akainic replied:
Nevermind, looks like `_.isPlainObject` won't return `true` for an array, that's good to know :)

akainic commented on spec/javascripts/health-care/utils/validations.spec.js:
```diff
@@ -105,4 +105,74 @@ describe('Validations unit tests', () => {
+          c: '',
+          d: 'str',
+          e: null,      // Ensure multiple null values are handled.
+          f: undefined  // Ensure undefined is not touched.
```
Can we also add a boolean condition? Since we're initializing checkboxes to all be `false` I want to make sure those aren't inadvertently changed.

akainic committed:
Fix child validations

awong-dev commented on _health-care/_js/components/HealthCareApp.jsx:
```diff
@@ -215,7 +214,17 @@ class HealthCareApp extends React.Component {
+      if (validations.isValidSection(this.props.location.pathname, sectionData)) {
+        hashHistory.push(this.getNextUrl());
+      } else {
+        document.getElementsByClassName('usa-input-error')[0].scrollIntoView();
```
Add a TODO here for me and maybe file an issue. I think this code may sometimes throw an exception.

akainic replied:
Yes, I noticed that, it's not the best :/ I'll add a TODO

awong-dev replied:
What's happening is `setState()` is not synchronous. See notes https://facebook.github.io/react/docs/component-api.html.  Multiple calls to setState() may get combined into one update to props.  Thus, the error class may not be set when this is called.

I don't know how to fix it "right" but to at least avoid the exception, check that `document.getElementsByClassName('usa-input-error')` isn't empty before trying to get element 0 and do nothing if it is empty. That will yield a bug where the window doesn't scroll but at least we won't throw an exception out of `setState()` which has undefined results.

awong-dev commented on _health-care/_js/utils/validations.js:
```diff
@@ -1,58 +1,207 @@
 }
 
+function isValidNameAndGeneralInformation(data) {
+  return this.isNotBlank(data.fullName.first) &&
```
The use of `this` here is weird.

It's probably working cause you're in a module and `this` is the module...but really, you should just be calling the function via name-lookup in lexical scoping. Drop all the `this.`?

awong-dev commented:
LGTM after one fix.

Reviewed...looks good enough to merge except for the `this.` usage in validations.js.

If there are any concerns, please just merge TBR with a TODO as this PR is getting too large to review.

akainic commented:
Fixed the `this` references in validations.js as well as the checking for if there are error classes before scrolling. Going to merge!
